components/calendar/src/helpers.rs

```diff
@@ -0,0 +1,69 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+/// Calculate `(n / d, n % d)` such that the remainder is always positive.
+pub fn div_rem_euclid(n: i32, d: i32) -> (i32, i32) {
+    debug_assert!(d > 0);
+    let (a, b) = (n / d, n % d);
+    if n >= 0 || b == 0 {
+        (a, b)
+    } else {
+        (a - 1, d + b)
+    }
```

Comment:
Explanation of how this works:

If the dividend is positive or if `d` exactly divides `n`, return `(n/d, n%d)` directly.

Otherwise, `n` is negative and the remainder is nonzero. We need to "fix up" the expression `n = ad + b` in order to make `b` positive. To do this, we set `a` to `a-1` and compensate by adding `d` to `b`.

Reply:
Note that `a - 1` won't overflow because the only way for `a` to be `i32::MIN` is for `d` to be `1`, and if this is the case, then `b == 0`.

Also, `d + b` can't overflow, because the result will always be inside in the open interval `(-d, d)`.